elrond-wasm/src/storage/mappers/token_attributes_mapper.rs

```diff
@@ -0,0 +1,239 @@
+
+pub struct TokenAttributesMapper<SA>
+where
+    SA: StorageReadApi + StorageWriteApi + ManagedTypeApi + ErrorApi + Clone + 'static,
+{
+    api: SA,
+    main_key: BoxedBytes,
+}
+
+impl<SA> StorageMapper<SA> for TokenAttributesMapper<SA>
+where
+    SA: StorageReadApi + StorageWriteApi + ManagedTypeApi + ErrorApi + Clone + 'static,
+{
+    fn new(api: SA, main_key: BoxedBytes) -> Self {
+        TokenAttributesMapper { api, main_key }
+    }
+}
+
+impl<SA> TokenAttributesMapper<SA>
+where
+    SA: StorageReadApi + StorageWriteApi + ManagedTypeApi + ErrorApi + Clone + 'static,
+{
+    pub fn set<T: elrond_codec::TopEncode + elrond_codec::TopDecode>(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+        attributes: &T,
+    ) {
+        let has_mapping = !self.is_empty_mapping_value(token_id);
+
+        let mapping = if has_mapping {
+            self.get_mapping_value(token_id)
+        } else {
+            let mut counter = self.get_counter_value();
+            if counter == u8::MAX {
+                self.api.signal_error(COUNTER_OVERFLOW_ERROR_MESSAGE);
+            }
+
+            counter += 1;
+            self.set_mapping_value(token_id, counter);
+            self.set_counter_value(counter);
+            counter
+        };
+
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
+        if has_value {
+            self.api.signal_error(VALUE_ALREADY_SET_ERROR_MESSAGE);
+        }
+
+        let mut value = Vec::<u8>::new();
+        attributes.top_encode(&mut value).unwrap();
+        self.set_token_attributes_value(mapping, token_nonce, value);
+    }
+
+    ///Use carefully. Update should be used mainly when backed up by the protocol.
+    pub fn update<T: elrond_codec::TopEncode + elrond_codec::TopDecode>(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+        attributes: &T,
+    ) {
+        let has_mapping = !self.is_empty_mapping_value(token_id);
+        if !has_mapping {
+            self.api.signal_error(UNKNOWN_TOKEN_ID_ERROR_MESSAGE);
+        }
+
+        let mapping = self.get_mapping_value(token_id);
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
+        if !has_value {
+            self.api
+                .signal_error(VALUE_NOT_PREVIOUSLY_SET_ERROR_MESSAGE);
+        }
+
+        let mut value = Vec::<u8>::new();
+        attributes.top_encode(&mut value).unwrap();
+        self.set_token_attributes_value(mapping, token_nonce, value)
+    }
+
+    pub fn clear(&self, token_id: &TokenIdentifier, token_nonce: u64) {
+        let has_mapping = !self.is_empty_mapping_value(token_id);
+        if !has_mapping {
+            self.api.signal_error(UNKNOWN_TOKEN_ID_ERROR_MESSAGE);
+        }
+
+        let mapping = self.get_mapping_value(token_id);
+        let has_value = !self.is_empty_token_attributes_value(mapping, token_nonce);
+        if !has_value {
+            self.api
```

Comment:
I think we should simply return here instead of signaling an error. That would be consistent with other mappers implementation as well.

Could be simply re-written as
if has_value {
self.clear_token_attributes_value(mapping, token_nonce);
}